### DISCUSSION TITLE
UCUM validation

### DISCUSSION THREAD

**Discussion by chgessner** [ucum-org/ucum#250] [Ideas]
This is a first draft for the definition of quality measures in the context of UCUM. The multi-layered logic of UCUM allows quality assessments on various levels. From an implementer perspective in many cases level A conformance is sufficient, given that appropriate value sets are provided. Level B addresses the quality of the value set, while level C looks into the functionality of tooling and converters.

A)	string comparison: This level does not relate to the UCUM essence or grammar at runtime, it works just like any other value-set based validation. The respective value set may be tailored to the intended use. The elements of the value set shall be validated against UCUM (validation levels B) at design time.

- A1)	literal: compare character by character 
- A2)	normalized representation: ignore leading and trailing whitespace
- A3)	literal without annotations: skip annotations while comparing

B)	grammar validation: This level is not usually employed at run-time. It is intended to qualify value sets and prepare for publication or production. It is based on a specific version of UCUM-esssence and corresponding syntax specifications.

- B1)	syntax check: check validity of term against UCUM-essence and grammar
- B2)	syntax-based conversion: attempt to convert to some “preferred representation”
- B3)	canonical representation: convert to representation that uses only terminal symbols 

C)	converter validation: this level of conformance addresses the conversion function of UCUM tooling. Aim is the correctness and precision of unit conversions. It is aimed at assessing the correctness of tools that are used for unit conversion, beyond validation of a single UCUM expression.

- C1)	constants: correctness of conversion factors and constants
- C2)	function definitions: correctness of “special units” (§§ 21, 22, 23 UCUM)

**Comment by dr-shorthair** [comment]
A2) normalization should probably convert `/` to `-n` or vice-versa as well?

**Comment by robmcclure** [comment]
I understand this proposal is from the terminology server perspective. This means:

- Level "A" means the server can only validate against existing (in the server) enumerated value sets based upon an enumerated set of UCUM codes as a representation of the code system. We need to clarify if this includes string validation against known annotations. 
- Level "B" means in addition to Level A validation, the server provides validation based upon understanding the UCUM grammar syntax.
- Level "C" means in addition to Level A and B, the server provides validation on units that UCUM can covert to or from.

**Comment by robmcclure** [comment]
I want to highlight that we do need to separately discuss the provision of annotations. UCUM does provide list of enumerated UCUM concepts with annotations included. In doing so we are providing some level on confirmation that they have some meaning to UCUM.
